packages/aragon-cli/src/lib/ipfs/index.js

```diff
@@ -0,0 +1,169 @@
+import chalk from 'chalk'
+import byteSize from 'byte-size'
+import { stringifyTree } from 'stringify-tree'
+import ipfsAPI from 'ipfs-http-client' // TODO: import only submodules?
+import fetch from 'node-fetch'
+
+const FETCH_TIMEOUT = 20000 // 20s
+const FETCH_TIMEOUT_ERR = 'Request timed out'
+
+export async function ensureConnection(rpc) {
+  try {
+    const client = connectToAPI(rpc)
+    await client.id()
+    return {
+      client,
+    }
+  } catch (e) {
+    throw new Error(`Could not connect to IPFS at ${JSON.stringify(rpc)}`)
+  }
+}
+
+export function connectToAPI(rpc) {
+  return ipfsAPI(rpc)
+}
+
+export async function getMerkleDAG(client, cid, opts = {}) {
+  const merkleDAG = parseMerkleDAG(await client.object.get(cid))
+  merkleDAG.cid = cid
+
+  if (opts.recursive && merkleDAG.isDir && merkleDAG.links) {
+    // fetch the MerkleDAG of each link recursively
+    const promises = merkleDAG.links.map(async link => {
+      const object = await getMerkleDAG(client, link.cid, opts)
+      return Object.assign(link, object)
+    })
+
+    return Promise.all(promises).then(links => {
+      merkleDAG.links = links
+      return merkleDAG
+    })
+  }
+
+  return merkleDAG
+}
+
+// object.get returns an object of type DAGNode
+// https://github.com/ipld/js-ipld-dag-pb#dagnode-instance-methods-and-properties
+function parseMerkleDAG(dagNode) {
+  const parsed = dagNode.toJSON()
+  // add relevant data
+  parsed.isDir = isDir(parsed.data)
+  // remove irrelevant data
+  delete parsed.data
+  if (!parsed.isDir) {
+    // if it's a big file it will have links to its other chunks
+    delete parsed.links
+  }
+  return parsed
+}
+
+function isDir(data) {
+  return data.length === 2 && data.toString() === '\u0008\u0001'
+}
+
+function stringifyMerkleDAGNode(merkleDAG) {
+  // ${merkleDAG.isDir ? '📁' : ''}
+  const cid = merkleDAG.cid
+  const name = merkleDAG.name || 'root'
```

Comment:
When viewing a `cid` that is a single file the name isn't display, only `root`. For example:
```
 aragon ipfs view QmNsZ1b2uXNEYKx17anDvRoJkvvmmeXovz7stKYDEbpRyW
 ✔ Connect to IPFS
 ✔ Fetch the links
─ root - 15.9kB - QmNsZ1b2uXNEYKx17anDvRoJkvvmmeXovz7stKYDEbpRyW
```
In this example the file is `artifact.json`. Ideally would be great to display the name.

Reply:
AFAIK the name of the hash cannot be retrieved. A way to preserve it would be to "wrap" files in a directory, that way, the DAG of the wrapper has a link to the file and it's name.

I think the idea behind it is that I can upload a file called `test.json` with the exact same contents of `artifact.json` and it will give me the same hash: `QmNsZ1b2uXNEYKx17anDvRoJkvvmmeXovz7stKYDEbpRyW` so the data is not duplicated or some other reason 🤔

Reply:
Oh, that's why 👌